lib/srv/alpnproxy/middleware.go

```diff
@@ -0,0 +1,145 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alpnproxy
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"net"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// LocalProxyMiddleware provides callback functions for LocalProxy.
+type LocalProxyMiddleware interface {
+	// OnNewConnection is a callback triggered when a new downstream connection is
+	// accepted by the local proxy.
+	OnNewConnection(ctx context.Context, lp *LocalProxy, conn net.Conn) error
+	// OnStart is a callback triggered when the local proxy starts.
+	OnStart(ctx context.Context, lp *LocalProxy) error
+}
+
+// DBCertChecker is a middleware that ensures that the local proxy has valid TLS database certs.
+type DBCertChecker struct {
+	// tc is a TeleportClient used to reissue certificates when necessary.
+	tc *client.TeleportClient
+	// dbRoute contains database routing information.
+	dbRoute tlsca.RouteToDatabase
+	// Clock specifies the time provider. Will be used to override the time anchor
+	// for TLS certificate verification.
+	// Defaults to real clock if unspecified
+	clock clockwork.Clock
+}
+
+func NewDBCertChecker(tc *client.TeleportClient, dbRoute tlsca.RouteToDatabase, clock clockwork.Clock) (LocalProxyMiddleware, error) {
+	if err := dbRoute.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if clock == nil {
+		clock = clockwork.NewRealClock()
+	}
+	return &DBCertChecker{
+		tc:      tc,
+		dbRoute: dbRoute,
+		clock:   clock,
+	}, nil
+}
+
+var _ LocalProxyMiddleware = (*DBCertChecker)(nil)
+
+// OnNewConnection is a callback triggered when a new downstream connection is
+// accepted by the local proxy.
+func (c *DBCertChecker) OnNewConnection(ctx context.Context, lp *LocalProxy, conn net.Conn) error {
+	return trace.Wrap(c.checkCerts(ctx, lp))
+}
+
+// OnStart is a callback triggered when the local proxy starts.
+func (c *DBCertChecker) OnStart(ctx context.Context, lp *LocalProxy) error {
+	return trace.Wrap(c.checkCerts(ctx, lp))
+}
+
+// needCertRenewal checks if the local proxy TLS certs are:
+// 1. configured
+// 2. not expired.
+func (c *DBCertChecker) needCertRenewal(lp *LocalProxy) (bool, error) {
+	certs := lp.GetCerts()
+	if len(certs) == 0 {
+		log.Debug("local proxy has no TLS certificates configured, need cert renewal")
+		return true, nil
+	}
+	err := utils.VerifyTLSCertificateExpiry(certs[0], c.clock)
+	if err != nil && errors.As(err, &x509.CertificateInvalidError{}) {
+		log.WithError(err).Debug("need cert renewal")
+		return true, nil
```

Comment:
nit. just curious if there are other reasons the certificate is invalid that may cause false positive here.

Reply:
I don't think any of them would be false positive, but there might be some path in `x509.ParseCertificate` that returns this error. I think I should actually trigger cert renewal for any error though, thanks for spotting this